src/main/java/in/projecteka/consentmanager/user/UserService.java

```diff
@@ -80,30 +80,55 @@
         if (!validateOtpVerification(otpVerification)) {
             throw new InvalidRequestException("invalid.request.body");
         }
-        return otpServiceClient
-                .verify(otpVerification.getSessionId(), otpVerification.getValue())
-                .then(signupService.generateToken(otpVerification.getSessionId()));
+
+        return signupService.getMobileNumber(otpVerification.getSessionId())
+                .switchIfEmpty(Mono.error(ClientError.networkServiceCallFailed()))
+                .flatMap(mobileNumber -> {
+                    OtpAttempt.OtpAttemptBuilder builder = OtpAttempt.builder()
+                            .sessionId(otpVerification.getSessionId())
+                            .identifierType("MOBILE")
+                            .identifierValue(mobileNumber)
+                            .action(OtpAttempt.Action.OTP_SUBMIT_REGISTRATION);
+                    return otpAttemptService.validateOTPSubmission(builder.build())
+                            .then(otpServiceClient.verify(otpVerification.getSessionId(), otpVerification.getValue(),
+                                    () -> otpAttemptService.saveOTPAttempt(builder.attemptStatus(OtpAttempt.AttemptStatus.FAILURE).build())))
+                            .then(otpAttemptService.removeMatchingAttempts(builder.build()))
+                            .then(signupService.generateToken(otpVerification.getSessionId()));
+                });
     }
 
     public Mono<Token> verifyOtp(OtpVerification otpVerification) {
         if (!validateOtpVerification(otpVerification)) {
             throw new InvalidRequestException("invalid.request.body");
         }
-        return otpServiceClient
-                .verify(otpVerification.getSessionId(), otpVerification.getValue())
-                .then(signupService.generateToken(new HashMap<>(), otpVerification.getSessionId()));
+        return signupService.getUserName(otpVerification.getSessionId())
+                .switchIfEmpty(Mono.error(ClientError.networkServiceCallFailed()))
+                .flatMap(userRepository::userWith)
+                .flatMap(user -> {
+                    OtpAttempt.OtpAttemptBuilder builder = OtpAttempt.builder()
+                            .sessionId(otpVerification.getSessionId())
+                            .identifierType("MOBILE")
+                            .identifierValue(user.getPhone())
+                            .action(OtpAttempt.Action.OTP_SUBMIT_RECOVER_PASSWORD)
+                            .cmId(user.getIdentifier());
+                    return otpAttemptService.validateOTPSubmission(builder.build())
+                            .then(otpServiceClient.verify(otpVerification.getSessionId(), otpVerification.getValue(),
+                                    () -> otpAttemptService.saveOTPAttempt(builder.attemptStatus(OtpAttempt.AttemptStatus.FAILURE).build())))
+                            .then(otpAttemptService.removeMatchingAttempts(builder.build()))
```

Comment:
some of the code seems duplicate for verifyOtp and permitOtp..can't we extract method out of it..I am not sure..just calling out loud.

Reply:
Yeah I feel there is duplication. But I think it's completely same. Because if one method we need CM ID and in other we just need the mobile number. So that's where I think it's hard to extract that out.

Let me know if you have any suggestions.